server/routers/status-page-router.js

```diff
@@ -80,20 +80,32 @@ router.get("/api/status-page/heartbeat/:slug", cache("1 minutes"), async (reques
         ]);
 
         for (let monitorID of monitorIDList) {
+            // Calculating the percentage of uptime per day with the status count to show full year data
             let list = await R.getAll(`
-                    SELECT * FROM heartbeat
+                    SELECT 
+                        strftime('%Y-%m-%d', time) AS day,
+                        (COUNT(CASE WHEN status = 1 THEN 1 END) * 100.0) / COUNT(status) AS percentage,
+                        (
+                            SELECT status
+                            FROM heartbeat AS sub
+                            WHERE strftime('%Y-%m-%d', sub.time) = strftime('%Y-%m-%d', main.time)
+                            GROUP BY status
+                            ORDER BY COUNT(*) DESC
+                            LIMIT 1
+                        ) as status
+                    FROM heartbeat as main
```

Comment:
Yes, in the upcoming version, the heartbeats table will only keep data for the last 48 hours. The `getDataArray` function in `UptimeCalculator` should be used instead. The raw counts data can retrieved that way, and the precentage calculated in the front-end.

 Since this was only implemented recently, with the only component utilizing it (`PingChart`) still in PR state, any feedback on the implementation is welcome.

Reply:
@chakflying correct me if i'm wrong, In near future ur gonna save only 48 hrs monitoring data right. so if I use `getDataArray` function, Will I get the last 365 days monitoring data?

Reply:
The `getDataArray` function does not use the `heartbeats` table, and instead uses the various `stats_` tables. You can take a look at the function implementation as it is in master. We have no time limit when using the 'day' interval.

```javascript
     getDataArray(num, type = "day") {
        if (type === "hour" && num > 24 * 30) {
            throw new Error("The maximum number of hours is 720");
        }
        if (type === "minute" && num > 24 * 60) {
            throw new Error("The maximum number of minutes is 1440");
        }
     ...

```